### PR TITLE
Update the CI

### PR DESCRIPTION
**Pull request by apomalyn**
In this PR we improve our CI.

Changes are the following:
- tagValidation: this job will check that the current version of the application isn't already deployed. In other words, if there is a tag (git tag) that equals the current version of the application (defined in the pubspec) the job will fail.
- validateWorkflow: test, analyze and check the format of the code. The tests are run with the coverage and a message will be posted to show the coverage of the application and the delta between your PR and the master branch (lcov.base.info). You do not need to push your lcov file, the CI will update all the file needed.
- android and iosBuildWorkflow: these aren't very useful for the moment but they build the ipa and apk file of the application. When we will be ready I will enable the auto-release which will use these file.

If you have any comment do not hesitate :smile:

### CODE REVIEW

**Comment by AlexandreMarkus** [comment]
I think the word should be version instead

**Review by AlexandreMarkus — COMMENTED** [review]
Very cool with the coverage. Can we add a coverage badge to the readme?

**Comment by apomalyn** [comment]
> I think the word should be version instead

Good catch! That was a test I didn't remove! :smile: 



> Very cool with the coverage. Can we add a coverage badge to the readme?

I'm still testing a way to add it dynamically. But yes we will have it soon! :smile:

File: .github/workflows/main-workflow.yaml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          lcov-file: ./coverage/lcov.info
+          lcov-base: ./coverage/lcov.base.info
+      - name: Commit coverage file
+        id: commit
+        if: ${{ github.event_name == 'push' }}
+        run: |
+          git config --local user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git config --local user.name "github-actions[bot]"
+          mv converage lcov.info lcov.base.info
```
[the remark this answers is not in the document]
Comment (apomalyn) [reply .github/workflows/main-workflow.yaml:94]: Yep...I switch my layout accidentally...

**Review by camillebrulotte — COMMENTED** [review]
Nice job ! :D
J'ai pas vraiment de commentaires à faire je m'y connais pas beaucoup en CI/fichier yaml...

**Comment by apomalyn** [comment]
> Nice job ! :D
> J'ai pas vraiment de commentaires à faire je m'y connais pas beaucoup en CI/fichier yaml...

T'inquiète c'est plus pour que vous faire découvrir le tout 😄  quand ce sera fonctionnel je ferais une petite présentation au club 😉

**Comment by AlexandreMarkus** [comment]
Maybe by the same type you could change the OpenSSL encryption algorithm, since it says that the one we are using is not safe. It also suggest two options in the warning message

**Comment by apomalyn** [comment]
> Maybe by the same time you could change the OpenSSL encryption algorithm, since it says that the one we are using is not safe. It also suggest two options in the warning message

This is because the initial file was encrypted using Mac with the default version of openssl (LibreSSL) who doesn't supports `-pbkdf2` or `-iter`. I just re-encrypted the file using the same algorithm with PBKDF2 and that should remove the warning!  😄

**Comment by apomalyn** [comment]
@ApplETS/general si des gens veulent regarder pourquoi le golden d'aboutView passe pas ^^'

**Comment by AlexandreMarkus** [comment]
> @ApplETS/general si des gens veulent regarder pourquoi le golden d'aboutView passe pas ^^'
> En local, si j'execute tout les tests ca fonctionne, cependant si j'execute le test seul il echoue --' le truc a regarder c'est si le golden genere possede bien le favicon du club ^^'

Ça fait quelques fois que ça se produit. J'ai remarqué qu'en relançant le workflow, la 2e fois ça passe.

**Comment by AlexandreMarkus** [comment]
> @ApplETS/general si des gens veulent regarder pourquoi le golden d'aboutView passe pas ^^'
> En local, si j'execute tout les tests ca fonctionne, cependant si j'execute le test seul il echoue --' le truc a regarder c'est si le golden genere possede bien le favicon du club ^^'

En regardant plus attentivement, c'est dû au même bug que #85 l'animation du hero lors de sa première apparition donne un résultat étrange.

**Comment by apomalyn** [comment]
@AlexandreMarkus c'est vraiment problématique comme bug si ca fait planté les tests XD

**Comment by AlexandreMarkus** [comment]
> @AlexandreMarkus c'est vraiment problématique comme bug si ca fait planté les tests XD

Effectivement!!

**Comment by apomalyn** [comment]
@AlexandreMarkus the image missing is loaded!! :smile: 

@ApplETS/general if someone ever encounters this problem, here is the [solution](https://github.com/flutter/flutter/issues/38997).

File: test/ui/views/more_view_test.dart [hunk 2]
```diff
@@ -131,6 +131,16 @@ void main() {
...
+                localizedWidget(child: FeatureDiscovery(child: MoreView())));
+            final Element element = tester.element(find.byType(Hero));
+            Hero widget = element.widget as Hero;
+            Image image = widget.child as Image;
+            await precacheImage(image.image, element);
+            await tester.pumpAndSettle();
+          });
+
           await tester.pumpWidget(
               localizedWidget(child: FeatureDiscovery(child: MoreView())));
```
Comment (AlexandreMarkus) [inline test/ui/views/more_view_test.dart]: I think you can remove this

Comment (AlexandreMarkus) [reply test/ui/views/more_view_test.dart]: dupplicate?

**Comment by apomalyn** [comment]
Will close #1

**Review by MysticFragilist — DISMISSED** [review]
LGTM 🍾 🎈  Really great work!

**Review by elornHamon — DISMISSED** [review]
Really loving the auto code coverage, LGTM ! 🚀

**Comment by apomalyn** [comment]
@MysticFragilist @elornHamon I still have some work to do (go check the project 😉 ) In this PR I would like to have the coverage badge, signing of the android application, release on Github 😄 

We will merge this PR when all the cards of the dashboard will be merged.

**Comment by MysticFragilist** [comment]
Ohh okay! tought it would be in another PR, I'll dismiss for now! (still a really good work Xavier!!)

**Comment by apomalyn** [comment]
@MysticFragilist at least I want to add the app signing 😄

**Comment by AlexandreMarkus** [comment]
@apomalyn let us know when it is ready for review

**Comment by apomalyn** [comment]
I have one last change for this PR. The app signing will be done later because we will probably sign the app using the Upload Keystore directly for Google so it is a special case.
The coverage badge will go to another PR also (except if I can finish it tonight).

So the only thing missing is a little bit of cleaning ^^

File: .github/workflows/main-workflow.yaml [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ matrix.target }}
+          path: $GITHUB_WORKSPACE/notre_dame_${{ matrix.target }}_${{ needs.tagValidation.outputs.version }}.${{ matrix.asset_extension }}
+
+  create-draft-release:
+    name: Draft Github release
+    if: ${{ github.event_name == 'push' }}
+    needs: [ tagValidation, validationWorkflow, create-build ]
+    runs-on: ubuntu-20.04
```
Comment (AlexandreMarkus) [inline .github/workflows/main-workflow.yaml:182]: Is this version needed? You could use Ubuntu latest

File: .github/workflows/release-workflow.yaml [hunk 4]
```diff
@@ -0,0 +1,193 @@
+name: Main workflow
```
Comment (AlexandreMarkus) [inline .github/workflows/release-workflow.yaml:1]: Maybe put a more accurate name : release workflow

**Review by AlexandreMarkus — COMMENTED** [review]
By the same time could you update the readme files with flutter 2.2 😛 my mistake

File: .github/workflows/release-workflow.yaml (hunk 4 above)
Comment (apomalyn) [reply .github/workflows/release-workflow.yaml:1]: This file will be deleted for now 😄  that was my mistake

File: .github/workflows/main-workflow.yaml (hunk 3 above)
Comment (apomalyn) [reply .github/workflows/main-workflow.yaml:182]: Yes, good catch!

**Comment by apomalyn** [comment]
@ApplETS/general this PR is ready for review :smile: (when the complete CI will work but the code and workflow will not be modified, only adjust to work)

**Review by AlexandreMarkus — DISMISSED** [review]
YEAH!!!!

**Review by camillebrulotte — DISMISSED** [review]
Good job ! :D

**Comment by apomalyn** [comment]
@AlexandreMarkus @camillebrulotte could you approve again the PR please? :sweat_smile: I will merge it right after :smile:

**Review by AlexandreMarkus — APPROVED** [review]

**Review by camillebrulotte — APPROVED** [review]

**Pull request merged**